feat: Auto-create Python classes when missing

junrushao opened the pull request.

junrushao commented on python/tvm_ffi/cython/object.pxi:
```diff
@@ -264,28 +266,42 @@ cdef inline object make_ret_opaque_object(TVMFFIAny result):
+        # in this case we need to report an warning
+
+        from .utils import prototype_py
+
```
Updated with an approach that doesn't rely on `exec`. Let me know if it makes sense

tqchen commented on python/tvm_ffi/cython/object.pxi:
```diff
@@ -261,29 +263,53 @@ cdef inline object make_ret_opaque_object(TVMFFIAny result):
+        # Slow path: object is not found in registered entry
+        # In this case we need to report a warning, create a dummy stub class
+        # and register the class for future usage.
+        cls = make_fallback_cls_for_type_index(type_index)
```
right, in this case documentation and naming clarity would help a bit

tqchen commented on python/tvm_ffi/cython/object.pxi:
```diff
@@ -261,29 +263,53 @@ cdef inline object make_ret_opaque_object(TVMFFIAny result):
+
+    warnings.warn(
+        f"Returning type `{type_key}` which is not registered via `@register_object`, "
+        f"falling back to auto-generated prototype:\n{type_info.prototype_py()}",
```
indeed it can be debatable. Originally I thought we need a warning.

However, now that we have a richer exposures of the class, it effectively behaves like pybind, and there could be effective usecases for such cxx only class in JITing scenarios, as a result the note. Alternatively, if we can have a context that marks this is intentional, which could also be fine. But i think no warning should also be ok

junrushao replied:
I see - good point! How about having warnings only if a certain env variable is turned on?

tqchen replied:
warning under a flag seems to be ok

junrushao commented on python/tvm_ffi/cython/type_info.pxi:
```diff
@@ -109,6 +110,50 @@ class TypeInfo:
+        """Generate Python source code for a prototype class of this type.
+
+        The generated class will have the correct inheritance structure and
+        field definitions, but will not have any method definitions.
```
It's not executed and is only used to provide hints to users who want to write `@c_class`. Should be pretty copy-paste-ready quality as it looks like:

```python
from typing import Any
from tvm_ffi import dataclasses as D
from tvm_ffi import core

PARENT_TYPE = core._lookup_type_info_from_type_key('ffi.Object').type_cls
assert PARENT_TYPE is not None

@D.c_class('testing.TestUnregisteredObject')
class TestUnregisteredObject(PARENT_TYPE):
  a: Any
  @staticmethod
  def __ffi_init__(*args): ...
```

tqchen replied:
Ah i see, i see the tention, i think ideally for the case where returning object is unregistered, maybe we do not need to make it a dataclass..

tqchen commented on python/tvm_ffi/cython/object.pxi:
```diff
@@ -261,29 +263,53 @@ cdef inline object make_ret_opaque_object(TVMFFIAny result):
+        f"falling back to auto-generated prototype:\n{type_info.prototype_py()}",
+    )
+
+    from tvm_ffi.dataclasses import c_class
```
I am not sure if the intention would be always the returning a dataclass in this case. My original thinking is that we could make it an Object with related fields and function being exported, unless user explicitly wanted a dataclass, in which case perhaps explicit `@c_class` decorator is a better idea.

junrushao commented:
Updated the PR per offline discussion with @tqchen. To summarize the changes:
- `prototype_py()` is no longer used in class creation;
- No warnings are thrown when encountering unregistered classes to be consistent with pybind behavior;
- `@c_class` is not used in this case, instead, we refactored existing `register_object(...)` behavior into `TypeMethod.as_callable(...)` and `TypeField.as_property(...)`, which is shared by both `register_object(...)` and `make_fallback_cls_for_type_index(...)`;
- Clarified in the doc that the slow path `make_fallback_cls_for_type_index` is called only once for every missing type.

CC: @tqchen

junrushao commented:
/gemini review

junrushao commented on python/tvm_ffi/cython/type_info.pxi:
```diff
@@ -109,6 +125,57 @@ class TypeInfo:
+        self.parent_type_info = _lookup_type_info_from_type_key(parent_type_key)  # ensure parent is registered
+
+    def prototype_py(self) -> str:
+        """Generate Python source code for a prototype class of this type.
```
sounds good

junrushao commented on python/tvm_ffi/cython/type_info.pxi:
```diff
@@ -101,6 +102,21 @@ class TypeMethod:
+        """Create a Python method attribute for this method on ``cls``."""
+        cdef str name = self.name
+        cdef str doc = self.doc or f"Method `{name}` of class `{cls.__qualname__}`"
+        cdef object func = self.func
+        if self.is_static:
+            func = staticmethod(func)
+        else:
+            func = _member_method_wrapper(func)
+        func.__module__ = cls.__module__
+        func.__name__ = name
+        func.__qualname__ = f"{cls.__qualname__}.{name}"
+        func.__doc__ = doc
+        return func
```
I tested locally but can't reproduce the claimed issue. @gemini Could you double check?

junrushao replied:
/gemini I tested locally but can't reproduce the claimed issue. Could you double check?

junrushao commented:
@gemini review

junrushao commented:
@tqchen Review comments addressed 🫡 Let me know if there's anything further I could iterate on!
